### ISSUE TITLE
DeepSpeed ignores trainer.gradient_accumulation

### ISSUE THREAD

**Issue by rob-hen** [Lightning-AI/pytorch-lightning#19350]
### Bug description

I am using the pytorch lightning CLI, and set `--trainer.strategy=deepspeed_stage_2`

By adding a print in `on_before_optimizer_step` hook, or `optimizer_step`, I can see the optimizer is called on each step, ignoring `trainer.accumulate_grad_batches`.

I am using pytorch lightning `2.0.9` and deepspeed `0.13.1`.

MWE (demo.py):
```python
from pytorch_lightning.cli import LightningCLI
from pytorch_lightning.demos.boring_classes import DemoModel, BoringDataModule

class Model(DemoModel):

    def optimizer_step(self, *args, **kwargs):
        print("CALLING OPTIMIZER")
        super().optimizer_step(*args, **kwargs)


def cli_main():
    cli = LightningCLI(Model, BoringDataModule)


if __name__ == "__main__":
    cli_main()

``` 
Call without deepspeed:
`python demo.py fit  --trainer.accumulate_grad_batches=32`
Call with deepspeed:
`python demo.py fit  --trainer.accumulate_grad_batches=32 --trainer.strategy=deepspeed_stage_2`

### What version are you seeing the problem on?

v2.0

### How to reproduce the bug

When running `python demo.py fit  --trainer.accumulate_grad_batches=32`, we get

![image](https://github.com/Lightning-AI/pytorch-lightning/assets/115473528/0edafaa9-8355-4763-8aeb-8af80e0ea654)
We can see the optimizer is called after 32 batches. 

When running `python demo.py fit  --trainer.accumulate_grad_batches=32 --trainer.strategy=deepspeed_stage_2`, we get
![image](https://github.com/Lightning-AI/pytorch-lightning/assets/115473528/e9c66da2-b529-4874-80c1-eb991a7a11a1)
We can see the optimizer is called after each batch. 

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

### More info

cc @awaelchli @Borda @carmocca @mauvilsa

**Comment by rob-hen** [comment]
Can it be that the pytorch lightning callback `optimizer_step ` is called on each step, but `super().optimizer_step(*args, **kwargs)` calls a DeepSpeed optimizer, that will correctly take care of gradient accumulation?

cc: @awaelchli

**Comment by awaelchli** [comment]
@rob-hen Yes, DeepSpeed handles gradient accumulation internally and for this reason Lightning needs to call into optimizer step for it to be handled there. The relevant lines of code are here in the trainer:

https://github.com/Lightning-AI/pytorch-lightning/blob/9d35c61f5f103b04b7fd1f504b8b937450482d07/src/lightning/pytorch/loops/optimization/automatic.py#L172-L176

Where `strategy.handles_gradient_accumulation == True` in the case of DeepSpeed, hence the "else" branch will be taken.

**Comment by awaelchli** [comment]
@rob-hen If you want to convince yourself that gradient accumulation is working you could for example mock/patch the optimizer.step() method and observe when it is getting called.

**Comment by rob-hen** [comment]
@awaelchli Thanks alot. I see. 

What about the other parameters that are mentioned in the [documentation](https://lightning.ai/docs/pytorch/stable/advanced/model_parallel/deepspeed.html#custom-deepspeed-config),
* optimizer
* learning rate
* scheduler

I define `learning rate`, `scheduler` and `optimizer` in `configure_optimizers`, how does it relate to defining these parameters in the configuration for DeepSpeed (as described in the documentation)?

Is the configuration for DeepSpeed automatically taken from PyTorch Lightning? Or I need to specify it again for DeepSpeed?

**Comment by awaelchli** [comment]
Defining the optimizer/scheduler in `configure_optimizers` is the preferred way. You don't need to include them in the config again.

**Comment by rob-hen** [comment]
`you could for example mock/patch the optimizer.step() method and observe when it is getting called` As far as I understood, `self._optimizer_step(batch_idx, closure)` is called on every step. So witj optimizer.step() you mean the function from DeepSpeed? Can you tell me where that function is?

**Comment by awaelchli** [comment]
```py
from torch.optim import SGD

class MySGD(SGD):
    def step(self, closure=None):
        print("optimizer step called")
        return super().step(closure)
        
        
# in LightningModule
def configure_optimizers(self):
    return MySGD(self.parameters(), lr=0.1)
```

**Comment by rob-hen** [comment]
So with **DeepSpeed** `optimizer_step` will be called every step, but step only after the actual gradient update, thus respecting the gradient accumulation?

**Comment by awaelchli** [comment]
Great!
